### PR TITLE
BUG: Use npy_log1p where appropriate in random generation

### PR DESCRIPTION
**Pull request by bashtage**
Extends #17020 and preserves legacy generators.

`log1p` was left in code where it makes no difference (e.g., `npy_log1pf(-next_double(...))` which is always the same as `log(1-next_double(...)`) since it seems more consistent.

### CODE REVIEW

**Comment by bashtage** [comment]
Only coverage failure, which I think is wrong here (but I can't see what codecov is seeing to check due to permissions).

File: numpy/random/_examples/cython/setup.py [hunk 1]
```diff
@@ -1,41 +1,43 @@
...
+    libraries=['npyrandom', 'npymath'],
+    define_macros=defs,
+)
 
 extensions = [extending, distributions]
 
-setup(
-    ext_modules=cythonize(extensions)
-)
+setup(ext_modules=cythonize(extensions))
```
Comment (mattip) [inline numpy/random/_examples/cython/setup.py:43]: These changes were needed?

**Review by mattip — APPROVED** [review]
Somehow codecoverage claims all the changes in distributions.c are not covered. Maybe because this is built as a library, and the optimization flags are somehow preventing the code coverage run?

In any case, this looks good to me.

File: numpy/random/_examples/cython/setup.py (hunk 1 above)
Comment (bashtage) [reply numpy/random/_examples/cython/setup.py:43]: A bit overagressive blackening probable.  I'll revert these.

Comment (bashtage) [reply numpy/random/_examples/cython/setup.py]: Reverted

**Comment by mattip** [comment]
Thanks @bashtage

**Pull request merged**